Add jflip module

cindywxw opened the pull request.

gfetterman commented:
There are three binary files in this PR. Can you modify `.gitignore` to remove them?

gfetterman commented:
There's a change to the top-level `SConstruct` file that seems to consist of the (unnecessary) addition of a single space. Including the change clutters the git history and the PR. Can you change it back?

gfetterman commented:
There's also a single-line change to `jill_module_skel.cc`. It isn't relevant to jflip, so please revert it.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+/*
```
This comment is no longer accurate - `jflip` isn't a skeleton any more. Please remove the first paragraph and the list and replace with a description of the module's function.

It's really important that documentation, including comments in source code, match the behavior of the program.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+#include <signal.h>
+#include <boost/shared_ptr.hpp>
+
+// #include <sys/time.h>
```
This can be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+#include "jill/jack_client.hh"
+#include "jill/program_options.hh"
+
+// https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/tree/master/2013/08/14/cpp
```
You aren't using the more complex randomization algorithm any more, so these includes should be removed, not commented out.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+    std::vector<string> input_ports;
+    /** A vector of outputs to connect to the client */
+    std::vector<string> output_ports;
+        /** The MIDI output channel */
```
`jflip` doesn't have a MIDI output port, so these lines can be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+
+static jflip_options options(PROGRAM_NAME);
+static boost::shared_ptr<jack_client> client;
+// boost::shared_ptr<dsp::crossing_trigger<sample_t> > trigger;
```
These commented lines aren't needed, so should be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+static int stopping = 0;
+static int xruns = 0;
+
+/* data storage for event times */
```
If this isn't needed, it should be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+    nframes_t inputSize = nframes;
+
+    memcpy(out, in, nframes * sizeof(sample_t));
+    /* rand() method */
```
This comment is probably unnecessary - it's clear enough from the loop code what's happening.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+    std::srand(std::time(NULL));
+    while( i < inputSize) { 
+        int coin = rand()%2;
+        if (coin == 0) *(out+i) *= -1;
```
This style can be easy to misread for anything more complex than `if (...) break` or `if (...) continue`. Can you bring the multiplication step down onto its own line?

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+        i++;
+    }
+
+    /* use Lemire's picking number method */
```
If you aren't using this method any more, it's better to remove it than comment it out.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+jack_xrun(jack_client *client, float delay)
+{
+        __sync_add_and_fetch(&xruns, 1); // gcc specific
+        // arf_thread->xrun();
```
`jflip` isn't writing to an ARF file, so this can be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+        __sync_add_and_fetch(&stopping, 1);
+        // wait for at least one process loop; not strictly async safe
+        usleep(2e6 * client->buffer_size() / client->sampling_rate());
+        // if (arf_thread) {
```
`jflip` isn't writing to an ARF file, so this can be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+signal_handler(int sig)
+{
+        ret = sig;
+        // if (arf_thread) {
```
`jflip` isn't writing to an ARF file, so this can be removed.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+                                                JackPortIsInput, 0);
+                port_out = client->register_port("out", JACK_DEFAULT_AUDIO_TYPE,
+                                                 JackPortIsOutput, 0);
+                // port_trig = client->register_port("trig_out",JACK_DEFAULT_MIDI_TYPE,
```
If this isn't needed, remove it.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+
+                client->set_buffer_size_callback(jack_bufsize);
+
+                // uncomment if you need these callbacks
```
If these aren't needed, remove them.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+
+                // activate client
+                client->activate();
+                // arf_thread->start();
```
This isn't needed; remove it.

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,291 @@
+                 "set client name")
+                ("in,i",      po::value<stringvec>(), "add connection to input port")
+                ("out,o",     po::value<stringvec>(), "add connection to output port")
+                // ("buffer",     po::value<float>(&buffer_size_s)->default_value(2.0), "minimum ringbuffer size (s)")
```
If this isn't needed, remove it.

cindywxw commented:
Thanks for your comments, Graham. I changed the jflip.cc and changed the lines in SConstruct and jill_module_skel.cc back. I'm not very familiar git commands so I'm not sure if I'm doing it right. Maybe you can help me with those files when we meet?

gfetterman commented on modules/jflip.cc:
```diff
@@ -0,0 +1,228 @@
+/*
+ * jflip - whiten the input data by randomly inverting it around
+ *
```
Can you add the license text back in (you can copy it from another module)?

gfetterman replied:
Also, the program description looks cut off. Randomly inverting it around what?

gfetterman commented:
Looks good. Thanks for making those changes.
